### PR TITLE
No branch protection metadata unless enabled

### PR DESCRIPTION
**Pull request by nagisa**
Even if we emit metadata disabling branch protection, this metadata may
conflict with other modules (e.g. during LTO) that have different branch
protection metadata set.

This is an unstable flag and feature, so ideally the flag not being
specified should act as if the feature wasn't implemented in the first
place.

Additionally this PR also ensures we emit an error if
`-Zbranch-protection` is set on targets other than the supported
aarch64. For now the error is being output from codegen, but ideally it
should be moved to earlier in the pipeline before stabilization.

### CODE REVIEW

File: src/test/codegen/branch-protection.rs [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
 
-// revisions: bti pac-ret leaf b-key
+// revisions: BTI PACRET LEAF BKEY NONE
 // min-llvm-version: 12.0.0
 // needs-llvm-components: aarch64
-// [bti] compile-flags: -Z branch-protection=bti
-// [pac-ret] compile-flags: -Z branch-protection=pac-ret
-// [leaf] compile-flags: -Z branch-protection=pac-ret,leaf
-// [b-key] compile-flags: -Z branch-protection=pac-ret,b-key
+// [BTI] compile-flags: -Z branch-protection=bti
```
Comment (jacobbramley) [inline src/test/codegen/branch-protection.rs:6]: Is it worth also testing the empty "`-Z branch-protection=`"? If I understand correctly, that should result in module flags being generated (with zero values), and that behaviour will be useful for validation before stabilisation.

**Comment by cjgillot** [comment]
Could you also add a test that branch protection does not run when not explicitly enabled?
Isn't changing this default a breaking change compared to some accidental insta-stabilization?

**Comment by nagisa** [comment]
> Could you also add a test that branch protection does not run when not explicitly enabled?

Isn't that covered by the `NONE` revision of the `branch-protection.rs` test?

File: src/test/codegen/branch-protection.rs (hunk 1 above)
Comment (nagisa) [reply src/test/codegen/branch-protection.rs:6]: Turns out no, empty value is not valid. Implementing that would be out of scope for this PR I feel.

**Comment by nagisa** [comment]
Nominating for beta since this missed the beta cut-off.

**Comment by Mark-Simulacrum** [comment]
(Note: I believe the beta being referenced here is 1.60, not 1.59 which is shipping later this week).

**Comment by apiraino** [comment]
Beta backport accepted (pending landing on nightly) as per compiler team [on Zulip](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-02-24.20.2354818.html#273101367)

@rustbot label +beta-accepted

**Pull request merged**